src/validation.cpp

```diff
@@ -1167,6 +1188,21 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
         // Of course, if an assumed valid block is invalid due to false scriptSigs
         // this optimization would allow an invalid chain to be accepted.
         if (fScriptChecks) {
+            // First check if script executions have been cached with the same
+            // flags. Note that this assumes that the inputs provided are
+            // correct (ie that the transaction hash which is in tx's prevouts
+            // properly commits to the scriptPubKey in the inputs view of that
+            // transaction).
+            uint256 hashCacheEntry;
+            // We only use the first 19 bytes of nonce to avoid a second SHA
+            // round - giving us 19 + 32 + 4 = 55 bytes (+ 8 + 1 = 64)
+            static_assert(55 - sizeof(flags) - 32 >= 128/8, "Want at least 128 bits of nonce for script execution cache");
+            CSHA256().Write(scriptExecutionCacheNonce.begin(), 55 - sizeof(flags) - 32).Write(tx.GetWitnessHash().begin(), 32).Write((unsigned char*)&flags, sizeof(flags)).Finalize(hashCacheEntry.begin());
+            AssertLockHeld(cs_main); //TODO: Remove this requirement by making CuckooCache not require external locks
+            if (scriptExecutionCache.contains(hashCacheEntry, !cacheFullScriptStore)) {
```

Comment:
> I'm confused, pvChecks is passed in, you dont populate, just push back your set of to-be-validated CScriptChecks?

I guess I wouldn't expect `cacheFullScriptStore` being set to true or false to have an effect on what gets returned in `pvChecks`, and that an interaction between the two parameters would be documented.

The specific case where it looks like there is an interaction is after calling `CheckInputs` once with `pvChecks = null` and a script where `check() == true` for all inputs.

If `CheckInputs` is called after that on the same script with `pvChecks != null`, it seems like `pvChecks` will be empty if `cacheFullScriptStore` was previously true, and populated if it was previously false.